### ISSUE TITLE
Virus Detection flagging library in syft for Mac install

### ISSUE THREAD

**Issue by bobpach** [anchore/syft#1753]
**What happened**:
Brew install stopped due to malware.

brew install syft
==> Downloading https://formulae.brew.sh/api/formula.jws.json
######################################################################## 100.0%
==> Downloading https://formulae.brew.sh/api/cask.jws.json
######################################################################## 100.0%
==> Fetching syft
==> Downloading https://ghcr.io/v2/homebrew/core/syft/manifests/0.78.0
Already downloaded: /Users/robertpacheco/Library/Caches/Homebrew/downloads/2988b3057bc1c524203d2c51dbd69883b4448bee5bb891dde376c94508e6344f--syft-0.78.0.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/syft/blobs/sha256:fa0b49e0437553c7033a09026014637a566ed06c017c2e5a609540c17d849173
Already downloaded: /Users/robertpacheco/Library/Caches/Homebrew/downloads/27680d977e5cff281c79e400090ff707a30e10103faeb17938b4906a1ea5c99f--syft--0.78.0.ventura.bottle.tar.gz
==> Pouring syft--0.78.0.ventura.bottle.tar.gz
cp: /private/tmp/d20230420-48576-hmuc07/syft/./0.78.0/bin/syft: Operation not permitted
Error: Failure while executing; `/usr/bin/env cp -pR /private/tmp/d20230420-48576-hmuc07/syft/. /usr/local/Cellar/syft` exited with 1. Here's the output:
cp: /private/tmp/d20230420-48576-hmuc07/syft/./0.78.0/bin/syft: Operation not permitted

OSX/Macstealer.ext found in soft/0.78.0/bin/syft.

**What you expected to happen**:

I expected a clean install

**Steps to reproduce the issue**:

On a Mac running Ventura 13.3.1 and Intego Virus Barrier 10.9.69

brew install syft

**Anything else we need to know?**:

**Environment**:
- Output of `syft version`: 0.78.0
- OS: Ventura 13.3.1
- Anti-virus:  Intego Virus Barrier 10.9.69

**Comment by bobpach** [comment]
Path correction: OSX/Macstealer.ext found in syft/0.78.0/bin/syft

**Comment by wagoodman** [comment]
Thanks for reporting, we're investigating now.

**Comment by joshbressers** [comment]
Hi @bobpach 

We're not able to reproduce this behavior. Are you able to give us a sha256sum of that syft binary?

**Comment by wagoodman** [comment]
Checking with the assets from the github release:
```
# source: https://github.com/anchore/syft/releases/tag/v0.78.0
❯ tree
.
├── amd
│   ├── syft
│   └── syft_0.78.0_darwin_amd64.tar.gz
├── arm
│   ├── syft
│   └── syft_0.78.0_darwin_arm64.tar.gz
└── syft_0.78.0_checksums.txt

❯ find . -type f -exec bash -c "sha256sum {};" \;
362736a28925a14fbbc126ce8f2bb3abd08e95280b0bcbc367c871f35b47b612  ./syft_0.78.0_checksums.txt
71406a97d307e0aaba1512b2a432a1bd8de2562f0c03082ac09242c6163f2917  ./amd/syft_0.78.0_darwin_amd64.tar.gz
b00e575220c6fcae3e4baaea57f9233800654970e180c7b19a0258066b720b82  ./amd/syft
1c23551a8120c7094abb7783f1a476a535616916010a538433dc4d3795761313  ./arm/syft_0.78.0_darwin_arm64.tar.gz
519574c28498dd909505d41dcc123c50ffa3c18497e588c3a8f59f2edd2d0be1  ./arm/syft
```

As well as the assets from homebrew core directly:
```
❯ crane manifest ghcr.io/homebrew/core/syft:0.78.0 | grep -C 10 "fa0b49e0437553c7033a09026014637a566ed06c017c2e5a609540c17d849173"
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:7886af280382429f3c454d0cd5acdf4c8fadef592728039d3a23717d06a473d2",
      "size": 1944,
      "platform": {
        "architecture": "amd64",
        "os": "darwin",
        "os.version": "macOS 13.3"
      },
      "annotations": {
        "org.opencontainers.image.ref.name": "0.78.0.ventura",
        "sh.brew.bottle.digest": "fa0b49e0437553c7033a09026014637a566ed06c017c2e5a609540c17d849173",
        "sh.brew.bottle.size": "13968063",
        "sh.brew.tab": "{\"homebrew_version\":\"4.0.14-31-gcf25397\",\"changed_files\":[],\"source_modified_time\":1681748767,\"compiler\":\"clang\",\"runtime_dependencies\":[],\"arch\":\"x86_64\",\"built_on\":{\"os\":\"Macintosh\",\"os_version\":\"macOS 13.3\",\"cpu_family\":\"penryn\",\"xcode\":\"14.3\",\"clt\":\"14.3.0.0.1.1679647830\",\"preferred_perl\":\"5.30\"}}"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:ac76eef7a64012b8978af1d3ec10e430dafc4c7fc41706e2e49e7ee1e9d8fac5",
      "size": 2029,
      "platform": {
        "architecture": "amd64",

❯ crane blob ghcr.io/homebrew/core/syft@sha256:fa0b49e0437553c7033a09026014637a566ed06c017c2e5a609540c17d849173 > syft.tar.gz

❯ tar -xzvf syft.tar.gz
...
x syft/0.78.0/bin/syft
...

❯ sha256sum syft/0.78.0/bin/syft
d42dc17f2adcc9827ab347bed899df27a3436d84f7abd830ea674ac8b9e8556e  syft/0.78.0/bin/syft

```

We uploaded these binaries to virus total all with clean scan results:
- github release amd64: https://www.virustotal.com/gui/file/b00e575220c6fcae3e4baaea57f9233800654970e180c7b19a0258066b720b82
- homebrew amd64: https://www.virustotal.com/gui/file/d42dc17f2adcc9827ab347bed899df27a3436d84f7abd830ea674ac8b9e8556e
- github release arm64: https://www.virustotal.com/gui/file/519574c28498dd909505d41dcc123c50ffa3c18497e588c3a8f59f2edd2d0be1 
- homebrew arm64: https://www.virustotal.com/gui/file/2571d34cc5e6d888ee8741f56909dc5093693d67eda83607b6bb2f6de180860a

**Comment by bobpach** [comment]
brew install --verbose syft
==> Fetching syft
==> Downloading https://ghcr.io/v2/homebrew/core/syft/manifests/0.78.0
Already downloaded: /Users/robertpacheco/Library/Caches/Homebrew/downloads/2988b3057bc1c524203d2c51dbd69883b4448bee5bb891dde376c94508e6344f--syft-0.78.0.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/syft/blobs/sha256:fa0b49e0437553c7033a09026014637a566ed06c017c2e5a609540c17d849173
Already downloaded: /Users/robertpacheco/Library/Caches/Homebrew/downloads/27680d977e5cff281c79e400090ff707a30e10103faeb17938b4906a1ea5c99f--syft--0.78.0.ventura.bottle.tar.gz
==> Verifying checksum for '27680d977e5cff281c79e400090ff707a30e10103faeb17938b4906a1ea5c99f--syft--0.78.0.ventura.bottle.tar.gz'
==> Pouring syft--0.78.0.ventura.bottle.tar.gz
/usr/bin/env tar --extract --no-same-owner --file /Users/robertpacheco/Library/Caches/Homebrew/downloads/27680d977e5cff281c79e400090ff707a30e10103faeb17938b4906a1ea5c99f--syft--0.78.0.ventura.bottle.tar.gz --directory /private/tmp/d20230420-68827-10bpu77
/usr/bin/env cp -pR /private/tmp/d20230420-68827-10bpu77/syft/. /usr/local/Cellar/syft
cp: /private/tmp/d20230420-68827-10bpu77/syft/./0.78.0/bin/syft: Operation not permitted
Error: Failure while executing; `/usr/bin/env cp -pR /private/tmp/d20230420-68827-10bpu77/syft/. /usr/local/Cellar/syft` exited with 1. Here's the output:
cp: /private/tmp/d20230420-68827-10bpu77/syft/./0.78.0/bin/syft: Operation not permitted

**Comment by bobpach** [comment]
I removed the tar file I had:
`rm /Users/robertpacheco/Library/Caches/Homebrew/downloads/27680d977e5cff281c79e400090ff707a30e10103faeb17938b4906a1ea5c99f--syft--0.78.0.ventura.bottle.tar.gz`

And then I retried the install with verbose:

brew install --verbose syft
==> Fetching syft
==> Downloading https://ghcr.io/v2/homebrew/core/syft/manifests/0.78.0
Already downloaded: /Users/robertpacheco/Library/Caches/Homebrew/downloads/2988b3057bc1c524203d2c51dbd69883b4448bee5bb891dde376c94508e6344f--syft-0.78.0.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/syft/blobs/sha256:fa0b49e0437553c7033a09026014637a566ed06c017c2e5a609540c17d849173
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:fa0b49e0437553c7033a09026014637a566ed06c017c2e5a609540c17d849173?se=2023-04-20T21%3A10%3A00Z&sig=hBMWUzm%2FJj0fG5W8TJih%2BuGVqKW9P4Po2tSf%2FCofV3I%3D&sp=r&spr=https&sr=b&sv=2019-12-12
/usr/bin/env /usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.14\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 13.3.1\)\ curl/7.87.0 --header Accept-Language:\ en --fail --retry 3 --location --remote-time --output /Users/robertpacheco/Library/Caches/Homebrew/downloads/27680d977e5cff281c79e400090ff707a30e10103faeb17938b4906a1ea5c99f--syft--0.78.0.ventura.bottle.tar.gz.incomplete https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:fa0b49e0437553c7033a09026014637a566ed06c017c2e5a609540c17d849173\?se=2023-04-20T21\%3A10\%3A00Z\&sig=hBMWUzm\%2FJj0fG5W8TJih\%2BuGVqKW9P4Po2tSf\%2FCofV3I\%3D\&sp=r\&spr=https\&sr=b\&sv=2019-12-12
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 13.3M  100 13.3M    0     0  19.6M      0 --:--:-- --:--:-- --:--:-- 19.8M
==> Verifying checksum for '27680d977e5cff281c79e400090ff707a30e10103faeb17938b4906a1ea5c99f--syft--0.78.0.ventura.bottle.tar.gz'
==> Pouring syft--0.78.0.ventura.bottle.tar.gz
/usr/bin/env tar --extract --no-same-owner --file /Users/robertpacheco/Library/Caches/Homebrew/downloads/27680d977e5cff281c79e400090ff707a30e10103faeb17938b4906a1ea5c99f--syft--0.78.0.ventura.bottle.tar.gz --directory /private/tmp/d20230420-69453-18wjg64
/usr/bin/env cp -pR /private/tmp/d20230420-69453-18wjg64/syft/. /usr/local/Cellar/syft
cp: /private/tmp/d20230420-69453-18wjg64/syft/./0.78.0/bin/syft: Operation not permitted
Error: Failure while executing; `/usr/bin/env cp -pR /private/tmp/d20230420-69453-18wjg64/syft/. /usr/local/Cellar/syft` exited with 1. Here's the output:
cp: /private/tmp/d20230420-69453-18wjg64/syft/./0.78.0/bin/syft: Operation not permitted

This is what I am seeing:

![Screenshot 2023-04-20 at 5 13 02 PM](https://user-images.githubusercontent.com/52288532/233488828-4b275e22-3636-4b8b-9073-c6e254dec054.png)

**Comment by joshbressers** [comment]
Thanks for the update @bobpach, we're working on getting to the bottom of this

**Comment by spiffcs** [comment]
Hey @bobpach! Just a quick update here - we've confirmed with the Intego Virus Barrier folks that this is a false positive - the signature used to detect the vulnerability was to broad and a correction should be issued soon.

We'll probably have a few more details to write up as a summary of the detection, investigation, and logging event before we close this issue, but just wanted to make sure we confirmed directly to you that this has been addressed and concluded as an FP at this time. Thanks again for the report!

**Comment by wagoodman** [comment]
@bobpach this was a fun one to dig into so I figured I'd share some details 😄 

When you first reported this issue we scanned the Syft binaries with the free version of Intego VirusBarrier and VirusTotal for extra measure, both scans were clean. We eventually figured out that the behavior you reported was only being triggered with the paid version of VirusBarrier, not the free version. With that clue in hand we wanted to figure next:

1. if this was a legitimate finding or a false positive?
2. how far back in our releases does VirusBarrier flag Syft releases?

When we discovered Intego has a malware submission page we submitted a copy of the suspicious Syft binary and also reached out to them directly to provide more context. While the Intego team was working on their side, we also started to look for more clues. 

After a lot more scans we found that it wasn't just the latest release of Syft that was being flagged (v0.78.0) but all releases starting from v0.47.0 (which was published in June 2022)! Focusing on the [release notes for v0.47.0](https://github.com/anchore/syft/releases/tag/v0.47.0) there were two new features that stood out:

> - Support newer versions of 'rpm' that use Sqlite for the db instead of BerkeleyDB #469
> - Support 'ndb' rpm database format used in rpmdb 4.15+ #504

These are both tied to a single PR, #1018, which swapped a dependency (github.com/knqyf263/go-rpmdb) from using our temporary fork back to the original upstream version. After a little experimentation and reading through PRs in `go-rpmdb` we were able to find a minimal reproducible snippet:

```golang
package main

import (
  "fmt"

  _ "modernc.org/sqlite"
)

func main() {
  fmt.Println("hello, world")
}
```

Which would be immediately flagged when running `CGO_ENABLED=false go build -o /tmp/test-build main.go`. 

Though `fmt.Println("hello, world")` looked _extremely_ suspicious (...jk...), the `modernc.org/sqlite` import was the sensitive line here. We only tried versions `v1.20.0`-`v1.22.0` of `modernc.org/sqlite` all of which were flagged by VirusBarrier.

`modernc.org/sqlite` is quite an interesting project, one I've been following for a little while now and was pretty gleeful when a new sqlite driver contender entered the ring, all without the need for `cgo`. Instead they transpiled the sqlite C code base into go using the `ccgo` tool (a little more info [here](https://datastation.multiprocess.io/blog/2022-05-12-sqlite-in-go-with-and-without-cgo.html) and [here](https://zserge.com/posts/transpilers/)... and the ["vulnerable" section](https://gitlab.com/cznic/sqlite/-/blob/master/lib/sqlite_darwin_amd64.go) seemed quite dense!). 

It was around this time the Intego team reached out and **confirmed that the finding against Syft was a false positive** and they would be pushing an update within the next day to curtail the existing signature for OSX/MacStealer so that Syft (and other projects) would not be incorrectly flagged by VirusBarrier. 🎉 🎉 _(and then there was much dancing)_

I'll leave this issue open for the meantime until the Intego VirusBarrier signature update is out and I can confirm that it no longer flags Syft. 

Thanks for reporting, @bobpach 👋
